### PR TITLE
api: rework Domain.is_registrable

### PR DESCRIPTION
**Pull request by peterthomassen**

### CODE REVIEW

**Comment by peterthomassen** [comment]
Ready for review. It's not clear to me why the API tests fail (they work locally).

File: api/desecapi/models.py [hunk 1]
```diff
@@ -259,19 +259,17 @@ def is_covered_by_foreign_zone(self):
         assert self.name == next(iter(private_domains), self.public_suffix)
 
         # Deny registration for non-local public suffixes and for domains covered by other users' zones
-        try:
-            owner = self.owner
-        except Domain.owner.RelatedObjectDoesNotExist:
-            owner = None
-        return Domain.objects.filter(Q(name__in=private_domains) & ~Q(owner=owner)).exists()
+        return Domain.objects.filter(Q(name__in=private_domains) & ~Q(owner=self.owner_or_none)).exists()
```
Comment (peterthomassen) [inline api/desecapi/models.py:261]: Alternative to this construction:

```
    def covers_foreign_zone(self):
        filter = Q(name__in=private_domains)
        try:
            filter &= ~Q(owner=self.owner)
        except Domain.owner.RelatedObjectDoesNotExist:
            pass
        return Domain.objects.filter(filter).exists()
```

The try-catch would have to be duplicated in `covers_foreign_zone`. Hm...

**Comment by peterthomassen** [comment]
Figured out the reason for the test failure, and fixed it. Ready from my perspective.

File: api/desecapi/models.py [hunk 2]
```diff
@@ -248,34 +236,59 @@ def is_registrable(self):
...
             owner = None
-        return not Domain.objects.filter(Q(name__in=private_domains) & ~Q(owner=owner)).exists()
+        return Domain.objects.filter(Q(name__in=private_domains) & ~Q(owner=owner)).exists()
+
+    def is_registrable(self):
+        """
+        Returns False in any of the following cases:
+        (a) the domain name is under .internal,
+        (b) the domain_name appears on the public suffix list,
+        (c) the domain is descendant to a zone that belongs to any user different from the given one,
```
[the remark this answers is not in the document]
Comment (nils-wisiol) [reply api/desecapi/models.py]: This is actually changed by a later commit, so nevermind

**Comment by peterthomassen** [comment]
> One thing that isn't included here yet is the `autoconfig.dedyn.io` case. Do we care? I did some research and could not find any standard document defining the autoconfig behavior, it seems to be Thunderbird-only and is only a security problem for Thunderbird users with @dedyn.io email addresses. So currently, it seems to not be a problem. But may it become in the future? Are there more hostnames like this?

`autodiscover.@` by Microsoft Cloudlook.

File: api/desecapi/models.py [hunk 3]
```diff
@@ -282,6 +282,10 @@ def is_registrable(self):
         if self.name == self.public_suffix and self.name not in settings.LOCAL_PUBLIC_SUFFIXES:
             return False
 
+        # Disallow _acme-challenge.dedyn.io and the like. Rejects direct children of public suffixes if prefixed by _
+        if self.name.count('.') == self.public_suffix.count('.') + 1 and self.name.startswith('_'):
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/models.py:285]: I don't think we need to accurately reflect each policy the decision in the docs. It increases docs maintenance work with negligible practical benefit (forbidden things stay forbidden anyways). The docs currently say that there may be policy limitations, and give some examples. I think that's fine.

File: api/desecapi/tests/test_domains.py [hunk 4]
```diff
@@ -54,19 +54,19 @@ def test_cant_register_global_non_local_public_suffix(self):
             global_public_suffixes=['com', 'de', 'xxx', 'com.uk'],
             local_public_suffixes=['something.else'],
         ):
+            self.assertNotRegistrable('tld')
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/tests/test_domains.py:57]: All TLDs are public suffixes by definition

File: api/desecapi/models.py [hunk 5]
```diff
@@ -248,34 +236,59 @@ def is_registrable(self):
...
+
+    def is_covered_by_foreign_zone(self):
         # Generate a list of all domains connecting this one and its public suffix.
         # If another user owns a zone with one of these names, then the requested
         # domain is unavailable because it is part of the other user's zone.
-        private_components = self.name.rsplit(public_suffix, 1)[0].rstrip('.')
+        private_components = self.name.rsplit(self.public_suffix, 1)[0].rstrip('.')
         private_components = private_components.split('.') if private_components else []
-        private_components += [public_suffix]
+        private_components += [self.public_suffix]
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/models.py:256]: I don't understand ...?

[the remark this answers is not in the document]
Comment (nils-wisiol) [reply api/desecapi/models.py:256]: `public_suffix` is not a "private component"

File: api/desecapi/models.py [hunk 6]
```diff
@@ -248,34 +236,59 @@ def is_registrable(self):
...
         private_domains = ['.'.join(private_components[i:]) for i in range(0, len(private_components) - 1)]
-        assert is_public_suffix or self.name == private_domains[0]
+        assert self.name == next(iter(private_domains), self.public_suffix)
 
         # Deny registration for non-local public suffixes and for domains covered by other users' zones
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/models.py:260]: done

File: api/desecapi/models.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/models.py:256]: slightly reworked the structure of the algorithm computing `private_domains` to avoid the issue itself

**Comment by peterthomassen** [comment]
I pushed a new commit to address your comments. If it looks good to you, I'll squash it.

Shall we block registering zone names `autodiscover` and `autoconfig`? It is already blocked if under a private domain owned by another user (and if the parent domain is not registered, then registering it will be blocked later with our new logic); the problem is therefore only for direct subzones of public suffixes, such as `autodiscover.s3.amazonaws.com` etc. What do you think?

**Comment by nils-wisiol** [comment]
Yeah let's have a hardcoded list of children of public suffixes that cannot be registered -- otherwise we may forget if we ever decide to add another local public suffix

**Comment by peterthomassen** [comment]
done! IMO the PR is ready to go. Let me know what you think, and I'll squash!

**Pull request merged**